### PR TITLE
#126 Href should use undecoded query for URI creation

### PR DESCRIPTION
**Pull request by dmzaytsev**
task #126
resolved issue with encoded query parts

### CODE REVIEW

**Comment by davvd** [comment]
Thanks for your pull request, let me find someone who can review it

**Comment by davvd** [comment]
@ggajos this pull request is for you,please review

File: src/test/java/org/takes/misc/HrefTest.java [hunk 1]
```diff
@@ -81,4 +81,17 @@ public void addsPath() throws IOException {
             Matchers.equalTo("http://example.com/e/f")
         );
     }
+
+    /**
+     * Href can accept encoded query part.
+     * @throws IOException If some problem inside
+     */
+    @Test
+    public void acceptsEncoded() throws IOException {
```
Comment (ggajos) [inline src/test/java/org/takes/misc/HrefTest.java]: @dmzaytsev let's name it `acceptsEncodedQuery`

File: src/test/java/org/takes/misc/HrefTest.java [hunk 2]
```diff
@@ -81,4 +81,17 @@ public void addsPath() throws IOException {
             Matchers.equalTo("http://example.com/e/f")
         );
     }
+
+    /**
+     * Href can accept encoded query part.
+     * @throws IOException If some problem inside
```
Comment (ggajos) [inline src/test/java/org/takes/misc/HrefTest.java]: @dmzaytsev let's add dot at the end of sentence.

File: src/main/java/org/takes/misc/Href.java [hunk 3]
```diff
@@ -64,7 +64,7 @@
     public Href(final String txt) {
         this.params = new ConcurrentHashMap<String, List<String>>(0);
         final URI link = URI.create(txt);
-        final String query = link.getQuery();
+        final String query = link.getRawQuery();
```
Comment (ggajos) [inline src/main/java/org/takes/misc/Href.java]: @dmzaytsev let's add also note in method comment how this method handles encoded queries

**Comment by ggajos** [comment]
@dmzaytsev all good, just few notes above

**Comment by ggajos** [comment]
@dmzaytsev we should add test also

**Comment by dmzaytsev** [comment]
@ggajos I added new method to HrefTest
Do you mean an another test ?

**Comment by dmzaytsev** [comment]
@ggajos I fix all but the last comment.
it's not clear what should be done,
explain please

**Comment by ggajos** [comment]
@dmzaytsev all good

**Comment by ggajos** [comment]
@rultor merge please

File: src/main/java/org/takes/misc/Href.java [hunk 4]
```diff
@@ -59,12 +59,13 @@
     /**
      * Ctor.
      * @param txt Text of the link
+     *  Href use undecoded query for URL creation
```
Comment (yegor256) [inline src/main/java/org/takes/misc/Href.java]: It's a strange place for a comment. Try to build project site with `mvn clean site -Psite` and open `./target/site/apidocs-1.0-SNAPSHOT/index.html`. Then, take a look how this particular method is documented, you will see the problem

**Comment by yegor256** [comment]
@dmzaytsev see my comment above

**Comment by dmzaytsev** [comment]
@yegor256 javadoc updated
although I'm not sure whether this comment is needed

**Comment by yegor256** [comment]
@dmzaytsev I also don't understand what this comment is actually giving our users. It should either be removed or re-written in a much verbose way, to actually explain what's going on here and why we're using this "raw" method. maybe give user some Java example?

**Comment by dmzaytsev** [comment]
@yegor256 I removed the comment

**Comment by yegor256** [comment]
@rultor merge

**Comment by dmzaytsev** [comment]
@yegor256 previous merge failed, could you repeat please?
two commits above - wrong branch, my mistake

**Comment by yegor256** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor merge

**Comment by davvd** [comment]
@elenavolokhova please, let us know what do you think about this ticket, according to our [QA rules](http://www.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor deploy pls

**Comment by elenavolokhova** [comment]
@ggajos
>davvd commented 7 days ago
Thanks for your pull request, let me find someone who can review it

 According to our [quality rules](http://www.teamed.io/qa.html):
 > Messages in a ticket always start with a name of a user they are addressed to.
 
 As far as you are the performer of this ticket you are responsible for its quality. Please make sure that everyone follows this rule in your tickets in the future. Ask them to fix issues and wait for fix before closing the review.

>dmzaytsev commented 4 days ago
@ggajos I fix all but the last comment.
it's not clear what should be done,
explain please

>ggajos commented 4 days ago
@dmzaytsev all good

According to our [quality rules](http://www.teamed.io/qa.html):
 > All problems found by the code reviewer were addressed by the pull request author.
 
 There was an issue found during review and not fixed. Make sure that all problems found in the code are fixed before closing the ticket next time.
 
 Please confirm that everything is clear and you have no questions regarding this approach.

**Comment by ggajos** [comment]
@elenavolokhova , is @davvd human?

**Comment by dmzaytsev** [comment]
@ggajos 
>The Matrix is everywhere. It is all around us

:)

**Comment by elenavolokhova** [comment]
@ggajos Am I human? In this project I start to doubt in this fact :) 
Anyway something should be done with @davvd's answers. As a responsible for quality in tickets you should ask to fix such issues.

In order to close this ticket I need your acknowledgment that everything is clear and you'll try to avoid all issues mentioned above.

**Comment by davvd** [comment]
@ggajos @dmzaytsev I fixed my comments above... humans :)

**Comment by ggajos** [comment]
@davvd sorry for doubts ;)

**Comment by ggajos** [comment]
@elenavolokhova agree, understand, clear

**Comment by ggajos** [comment]
@dmzaytsev it is :)

**Comment by elenavolokhova** [comment]
@davvd Ticket finally passed QA review :) Quality is acceptable here.

**Comment by davvd** [comment]
> @davvd Ticket finally passed QA review :) Quality is acceptable here.

@elenavolokhova thanks, got it. everybody, please try to make it good next time

**Comment by davvd** [comment]
@ggajos just added 10 mins to @elenavolokhova (for QA), payment ID is `55489942`. paid, thanks, added **22 mins** to your account, payment `AP-5KE78986PJ115081L`, 111 hours and 34 mins was spent. review comments (c=7) added as a bonus. +22 added to your rating, current score is: [+3700](http://www.netbout.com/b/35094?open=rating)

**Pull request merged**